cmd/heimdalld/main.go

```diff
@@ -138,6 +184,254 @@ func exportAppStateAndTMValidators(logger log.Logger, db dbm.DB, storeTracer io.
+	addr := viper.GetString(flagAddress)
+	home := viper.GetString("home")
+	traceWriterFile := viper.GetString(flagTraceStore)
+
+	db, err := openDB(home)
+	if err != nil {
+		return err
+	}
+	traceWriter, err := openTraceWriter(traceWriterFile)
+	if err != nil {
+		return err
+	}
+
+	app := appCreator(ctx.Logger, db, traceWriter)
+
+	svr, err := tserver.NewServer(addr, "socket", app)
+	if err != nil {
+		return fmt.Errorf("error creating listener: %v", err)
+	}
+
+	svr.SetLogger(ctx.Logger.With("module", "abci-server"))
+
+	err = svr.Start()
+	if err != nil {
+		common.Exit(err.Error())
+	}
+
+	common.TrapSignal(ctx.Logger, func() {
+		// cleanup
+		err = svr.Stop()
+		if err != nil {
+			common.Exit(err.Error())
+		}
+	})
+
+	// run forever (the node will not be returned)
+	select {}
+}
+
+func startInProcess(ctx *server.Context, appCreator server.AppCreator, cdc *codec.Codec, startRestServer bool, startBridge bool) (*node.Node, error) {
+	cfg := ctx.Config
+	home := cfg.RootDir
+	traceWriterFile := viper.GetString(flagTraceStore)
+
+	db, err := openDB(home)
+	if err != nil {
+		return nil, err
+	}
+	traceWriter, err := openTraceWriter(traceWriterFile)
+	if err != nil {
+		return nil, err
+	}
+
+	app := appCreator(ctx.Logger, db, traceWriter)
+
+	nodeKey, err := p2p.LoadOrGenNodeKey(cfg.NodeKeyFile())
+	if err != nil {
+		return nil, err
+	}
+
+	server.UpgradeOldPrivValFile(cfg)
+
+	// create & start tendermint node
+	tmNode, err := node.NewNode(
+		cfg,
+		pvm.LoadOrGenFilePV(cfg.PrivValidatorKeyFile(), cfg.PrivValidatorStateFile()),
+		nodeKey,
+		proxy.NewLocalClientCreator(app),
+		node.DefaultGenesisDocProviderFunc(cfg),
+		node.DefaultDBProvider,
+		node.DefaultMetricsProvider(cfg.Instrumentation),
+		ctx.Logger.With("module", "node"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := tmNode.Start(); err != nil {
+		return nil, err
+	}
+
+	var cpuProfileCleanup func()
+
+	if cpuProfile := viper.GetString(flagCPUProfile); cpuProfile != "" {
+		f, err := os.Create(cpuProfile)
+		if err != nil {
+			return nil, err
+		}
+
+		ctx.Logger.Info("starting CPU profiler", "profile", cpuProfile)
+		if err := pprof.StartCPUProfile(f); err != nil {
+			return nil, err
+		}
+
+		cpuProfileCleanup = func() {
+			ctx.Logger.Info("stopping CPU profiler", "profile", cpuProfile)
+			pprof.StopCPUProfile()
+			f.Close()
+		}
+	}
+
+	// start rest
+	if startRestServer {
+		restCh := make(chan struct{})
+		go func() {
+			_ = restServer.StartRestServer(cdc, hmserver.RegisterRoutes, restCh)
+		}()
+		<-restCh
+	}
+
+	// start bridge
+	if startBridge {
+		go func() {
+			bridgeCmd.StartBridge(false)
+		}()
+	}
+
+	server.TrapSignal(func() {
+		ctx.Logger.Info("trap signal")
+
+		if tmNode.IsRunning() {
+			_ = tmNode.Stop()
+		}
+
+		if cpuProfileCleanup != nil {
+			cpuProfileCleanup()
+		}
+
+		ctx.Logger.Info("exiting...")
+	})
+
+	// run forever (the node will not be returned)
```

Comment:
Is this the thing you mentioned about graceful shutdown right? Please add a TODO.

Reply:
yes, I will add, this function is blocking, and we will have to change the logic in server.TrapSignal so that there is no os.Exit since it will prevent gracefully shutdown of the rest server, bridge.

Reply:
yes, I will add, this function is blocking, and we will have to change the logic in server.TrapSignal so that there is no os.Exit since it will prevent gracefully shutdown of the rest server, bridge.